Adopt C++20 Concepts in JavaScriptCore/dfg

iidmsa opened the pull request:
#### 56ff086f9da348f91908594db9e741485c043863
<pre>
Adopt C++20 Concepts in JavaScriptCore/dfg
<a href="https://bugs.webkit.org/show_bug.cgi?id=302683">https://bugs.webkit.org/show_bug.cgi?id=302683</a>
<a href="https://rdar.apple.com/164931352">rdar://164931352</a>

Reviewed by Darin Adler and Sam Weinig.

* Source/JavaScriptCore/dfg/DFGCFG.h:
* Source/JavaScriptCore/dfg/DFGDominators.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::queriedType):
(JSC::DFG::Node::OpInfoWrapper::as const):
* Source/JavaScriptCore/dfg/DFGOSRExit.cpp:
(JSC::DFG::OSRExit::emitRestoreArguments):
* Source/JavaScriptCore/dfg/DFGOpInfo.h:
* Source/JavaScriptCore/dfg/DFGSlowPathGenerator.h:
(JSC::DFG::CallSlowPathGenerator::tearDown):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
(JSC::DFG::SpeculativeJIT::tryHandleOrGetExceptionUnderSilentSpill):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileMapGetImpl):

Canonical link: <a href="https://commits.webkit.org/303328@main">https://commits.webkit.org/303328@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e39e9aa104770f5d410e7638e288d2386e95a3b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131973 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4465 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42987 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139487 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83871 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4227 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100877 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134919 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3144 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118191 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81668 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3038 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/892 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82705 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124036 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111791 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36315 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142131 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130480 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4135 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36895 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109248 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4216 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3604 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109418 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27730 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3128 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114471 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57356 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4188 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32868 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163447 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4020 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67635 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42504 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4280 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4148 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/dfg/DFGNode.h:
```diff
@@ -1659,7 +1659,7 @@ struct Node {
     JSType queriedType()
     {
-        static_assert(std::is_same<uint8_t, std::underlying_type<JSType>::type>::value, "Ensure that uint8_t is the underlying type for JSType.");
+        static_assert(std::same_as<uint8_t, std::underlying_type_t<JSType>>, "Ensure that uint8_t is the underlying type for JSType.");
```
```suggestion
        static_assert(std::same_as<uint8_t, std::underlying_type_t<JSType>>);
```

weinig commented on Source/JavaScriptCore/dfg/DFGCFG.h:
```diff
@@ -96,13 +96,15 @@ class CPSCFG : public SingleRootGraph<CFG> {
-template <typename T, typename = typename std::enable_if<std::is_same<T, CPSCFG>::value>::type>
+template<typename T>
+    requires std::same_as<T, CPSCFG>
 CPSCFG& selectCFG(Graph& graph)
```
This is good. Just wanted to through out two alternative mechanism we could use incase they seem nicer in anyway.

struct specialization:
```c++
template<typename> struct CFGSelection;
template<> struct CFGSelection<CPSCFG> {
    static CPSCFG& selectCFG(Graph& graph)
    { 
        return graph.ensureCPSCFG(); 
    }
};
template<> struct CFGSelection<SSACFG> {
    static SSACFG& selectCFG(Graph& graph)
    { 
        RELEASE_ASSERT(graph.m_ssaCFG);
        return *graph.m_ssaCFG;
    }
};

template<typename T>
auto& selectCFG(Graph& graph)
{
    return SelectCFG<T>::select(graph);
}
```

if constexpr:
```c++
template<typename T>
auto& selectCFG(Graph& graph)
{
    if constexpr (std::same_as<T, CPSCFG>) {
        return graph.ensureCPSCFG();
    } else if constexpr (std::same_as<T, SSACFG>) {
        RELEASE_ASSERT(graph.m_ssaCFG);
        return *graph.m_ssaCFG;
    }
}
```

weinig replied:
In a followup, we should also probably add LIFETIME_BOUND to these.

iidmsa replied:
> This is good. Just wanted to through out two alternative mechanism we could use incase they seem nicer in anyway.
> 
> struct specialization:
> 
> ```c++
> template<typename> struct CFGSelection;
> template<> struct CFGSelection<CPSCFG> {
>     static CPSCFG& selectCFG(Graph& graph)
>     { 
>         return graph.ensureCPSCFG(); 
>     }
> };
> template<> struct CFGSelection<SSACFG> {
>     static SSACFG& selectCFG(Graph& graph)
>     { 
>         RELEASE_ASSERT(graph.m_ssaCFG);
>         return *graph.m_ssaCFG;
>     }
> };
> 
> template<typename T>
> auto& selectCFG(Graph& graph)
> {
>     return SelectCFG<T>::select(graph);
> }
> ```
> 
> if constexpr:
> 
> ```c++
> template<typename T>
> auto& selectCFG(Graph& graph)
> {
>     if constexpr (std::same_as<T, CPSCFG>) {
>         return graph.ensureCPSCFG();
>     } else if constexpr (std::same_as<T, SSACFG>) {
>         RELEASE_ASSERT(graph.m_ssaCFG);
>         return *graph.m_ssaCFG;
>     }
> }
> ```

Resorted to Struct Specialization.

weinig commented on Source/JavaScriptCore/dfg/DFGOpInfo.h:
```diff
@@ -41,9 +41,8 @@ namespace JSC { namespace DFG {
-        typename IntegralType,
-        typename Constraint = typename std::enable_if<(std::is_integral<IntegralType>::value || std::is_enum<IntegralType>::value) && sizeof(IntegralType) <= sizeof(uint64_t)>::type>
+    template<typename IntegralType>
+        requires ((std::integral<IntegralType> || std::is_enum_v<IntegralType>) && sizeof(IntegralType) <= sizeof(uint64_t))
```
In a followup, adding a WTF::IntegralOrEnum concept might be a nice cleanup.

iidmsa replied:
Sounds good!

iidmsa committed:
Adopt C++20 Concepts in JavaScriptCore/dfg
https://bugs.webkit.org/show_bug.cgi?id=302683
rdar://164931352

Reviewed by Darin Adler and Sam Weinig.

* Source/JavaScriptCore/dfg/DFGCFG.h:
* Source/JavaScriptCore/dfg/DFGDominators.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::queriedType):
(JSC::DFG::Node::OpInfoWrapper::as const):
* Source/JavaScriptCore/dfg/DFGOSRExit.cpp:
(JSC::DFG::OSRExit::emitRestoreArguments):
* Source/JavaScriptCore/dfg/DFGOpInfo.h:
* Source/JavaScriptCore/dfg/DFGSlowPathGenerator.h:
(JSC::DFG::CallSlowPathGenerator::tearDown):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
(JSC::DFG::SpeculativeJIT::tryHandleOrGetExceptionUnderSilentSpill):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileMapGetImpl):

Canonical link: https://commits.webkit.org/303328@main
